### PR TITLE
normalizing flow networks

### PR DESCRIPTION
**Pull request by hnyu**
Implemented a normalizing flow realNVP in issue #1198 

See the docstring screenshot below for a brief introduction to realNVP. So far we assume a flattened rank-1 tensor to be transformed by realNVP. However, the conditional input can be very complex.

image

Also based on this, implemented a LatentActorDistributionNetwork class which can wrap any actor distribution network with a normalizing flow network. 

I tried defining a latent actor distribution network used by SAC and ran it on Pendulum, and got reasonable performance: 

sac_pendulum_latent_actor

### CODE REVIEW

File: alf/networks/normalizing_flow_networks.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+        if self._conditional_inputs:
+            y, z = yz
+        else:
+            y, z = yz, None
+        transform = self.make_invertible_transform(z)
+        return transform.inv(y), ()
+
+
+@alf.configurable
+class realNVPNetwork(NormalizingFlowNetwork):
```
Comment (emailweixu) [inline alf/networks/normalizing_flow_networks.py]: real => Real ?

File: alf/networks/normalizing_flow_networks.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+        r"""Express the network forward computation as an invertible PyTorch
+        ``Transform``. This overall transformation can be a composed one chaining
+        many transformation layers.
+
+        Args:
+            conditional_inputs: an optional nested conditional inputs that
+                condition the mapping :math:`x \rightarrow y`.
+
+        Returns:
+            torch.td.Transform: an invertible transform
```
Comment (emailweixu) [inline alf/networks/normalizing_flow_networks.py]: return type can be annotated: "def func_name() -> return_type:"

File: alf/networks/normalizing_flow_networks.py [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+        """
+        if not self._use_transform_cache:
+            return self._make_invertible_transform(conditional_inputs)
+        old_t, old_z = self._cached_transform
+        if conditional_inputs is old_z and old_t is not None:
```
Comment (emailweixu) [inline alf/networks/normalizing_flow_networks.py:105]: How is it possible that conditional_inputs is old_z and old_t is None?

File: alf/networks/normalizing_flow_networks.py [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+            return old_t
+        t = self._make_invertible_transform(conditional_inputs)
+        self._cached_transform = (t, conditional_inputs)
+        return t
+
+    def _make_invertible_transform(
+            self, conditional_inputs: alf.nest.NestedTensor = None):
```
Comment (emailweixu) [inline alf/networks/normalizing_flow_networks.py:112]: This function should be dcumented since subclass needs to implement this function.

**Review by emailweixu — COMMENTED** [review]
One general issue is that it is not compatible with: `extract_spec()`, `distributions_to_params` and `params_to_distributions()`.
For example, in RLAlgorithm.unroll(), rollout_info_spec is extracted from the result of one rollout_step(), where the Transform object is stored in the spec. And the same Transform object is used when params_to_distributions() is used. But in the case of NF, different Transform object is created at different rollout_step.

Similar problem for `Algorithm_collect_train_info_sequentially()`

**Comment by hnyu** [comment]
> One general issue is that it is not compatible with: `extract_spec()`, `distributions_to_params` and `params_to_distributions()`. For example, in RLAlgorithm.unroll(), rollout_info_spec is extracted from the result of one rollout_step(), where the Transform object is stored in the spec. And the same Transform object is used when params_to_distributions() is used. But in the case of NF, different Transform object is created at different rollout_step.
> 
> Similar problem for `Algorithm_collect_train_info_sequentially()`

Right, the problem is that currently the transform objects are always using the conditional inputs at the first step, which is equivalent to not condition on anything.

**Comment by emailweixu** [comment]
> > One general issue is that it is not compatible with: `extract_spec()`, `distributions_to_params` and `params_to_distributions()`. For example, in RLAlgorithm.unroll(), rollout_info_spec is extracted from the result of one rollout_step(), where the Transform object is stored in the spec. And the same Transform object is used when params_to_distributions() is used. But in the case of NF, different Transform object is created at different rollout_step.
> > Similar problem for `Algorithm_collect_train_info_sequentially()`
> 
> Right, the problem is that currently the transform objects are always using the conditional inputs at the first step, which is equivalent to not condition on anything.

`distributions_to_params` needs to be modified to extract the conditional inputs as params.

File: alf/networks/normalizing_flow_networks.py (hunk 3 above)
Comment (hnyu) [reply alf/networks/normalizing_flow_networks.py:99]: When conditional_inputs is None and the default value of old_z is also None

Comment (hnyu) [reply alf/networks/normalizing_flow_networks.py:105]: > How is it possible that conditional_inputs is old_z and old_t is None?

When conditional_inputs is None and the default value of old_z is also None

File: alf/networks/normalizing_flow_networks.py (hunk 1 above)
Comment (hnyu) [reply alf/networks/normalizing_flow_networks.py:155]: > real => Real ?

'realNVP' is what the paper uses to call their method.

**Comment by Haichao-Zhang** [comment]
> maybe we can change the implementation of AC to compute all probs-related values in rollout_step()?

We already support prob computation at rollout_step: #1055 ?

File: alf/networks/normalizing_flow_networks.py (hunk 4 above)
Comment (emailweixu) [reply alf/networks/normalizing_flow_networks.py:112]: This is an interface which the subclass needs to implement. So its expected behavior should be documented.

File: alf/networks/normalizing_flow_networks.py [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+        xy: torch.Tensor,
+        z_spec: alf.nest.NestedTensorSpec = None,
+        z: alf.nest.NestedTensor = None
+) -> Tuple[alf.nest.NestedTensor, alf.utils.tensor_utils.BatchSquash]:
+    """A general function for adjusting the shapes of inputs and conditional inputs
+    of a conditional flow, prepared for a forward of a network next. Some networks
+    assume only one batch dim, for example, when using ``alf.layers.Reshape()``.
+
+    The reason why we need to do this is because the flow transform can be called
+    with an arbitrary batch shape of ``x`` or ``y``, for example, with computing
```
Comment (emailweixu) [inline alf/networks/normalizing_flow_networks.py]: "with computing" => "when computing"

File: alf/networks/normalizing_flow_networks.py [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+    a loss with time dimension, or sampling a particular shape from a distribution.
+
+    Args:
+        xy_spec: tensor spec of ``x`` (forward) or ``y`` (inverse)
+        xy:
+        z_spc: tensor spec of ``z`` (conditional input)
```
Comment (emailweixu) [inline alf/networks/normalizing_flow_networks.py]: z_spc => z_spec

File: alf/networks/normalizing_flow_networks.py [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+
+        if z_outer_rank > 1:
+            z = alf.utils.tensor_utils.BatchSquash(z_outer_rank).flatten(z)
+
+        B = alf.nest.get_nest_batch_size(z)
+        if B < ret.shape[0]:
+            # When the total outer dim of ``z`` is smaller than that of ``xy``,
+            # it means that multiple samples of ``xy`` correspond to one ``z``,
+            # so we need to repeat ``z``'s batch dim.
+            assert ret.shape[0] % B == 0, (
```
Comment (emailweixu) [inline alf/networks/normalizing_flow_networks.py]: Isn't the assert at line 398 enough to ensure this?

File: alf/networks/normalizing_flow_networks.py [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+        new_y = (y - trans) / self._scale_nonlinear(scale)
+        x = y * self._b + new_y * (~self._b)
+        return x
+
+    def log_abs_det_jacobian(self, x, y):
+        r"""The Jacobian is always a triangular matrix (or can be converted into by
+        row swapping). The diagonal elements are :math:`\mathbb{I}_d` and
+        :math:`\text{diag}(\exp(scale(x_{1:d};z)))`, where the first :math:`d` dims
+        are assumed to be selected by the mask ``self._b``.
+        """
```
Comment (emailweixu) [inline alf/networks/normalizing_flow_networks.py:532]: From the name of this function, it calculates log(det (Jacobian)), which should be a scalar quantity (with appropriate batch dimension). But the implementation is not so.

Comment (hnyu) [reply alf/networks/normalizing_flow_networks.py:532]: > From the name of this function, it calculates log(det (Jacobian)), which should be a scalar quantity (with appropriate batch dimension). But the implementation is not so.

This is just the name convention of PyTorch. For other transforms like StableTanh the returned value of this function can also be a vector. PyTorch will handle this smartly when compute ``log_prob()`` for a transformed distribution.

File: alf/networks/normalizing_flow_networks.py [hunk 9]
```diff
@@ -0,0 +1,538 @@
...
+        ret = (ret, z)
+
+    return ret, bs
+
+
+class _RealNVPTransform(td.Transform):
+    """This class implements each transformation layer of ``RealNVPNetwork``. For
+    details, refer to the docstring of ``RealNVPNetwork``.
+    """
+    domain = td.constraints.real
```
Comment (emailweixu) [inline alf/networks/normalizing_flow_networks.py]: domain and codomain should be td.constraints.real_vector

File: alf/networks/normalizing_flow_networks.py (hunk 8 above)
Comment (emailweixu) [reply alf/networks/normalizing_flow_networks.py:532]: It depends on its domain. If the domain is real, it a transform over a single real value and the returned Tensor has same shape as x. But if the domain is real_vector, the ndim of the returned Tensor should 1 less than that of x.

Comment (hnyu) [reply alf/networks/normalizing_flow_networks.py:532]: > It depends on its domain. If the domain is real, it a transform over a single real value and the returned Tensor has same shape as x. But if the domain is real_vector, the ndim of the returned Tensor should 1 less than that of x.

Right. It does depend on the domain. Currently I have real scalar for RealNVPTransform. But I think there is no difference between the final probs of the two choices?

Comment (hnyu) [reply alf/networks/normalizing_flow_networks.py:532]: > It depends on its domain. If the domain is real, it a transform over a single real value and the returned Tensor has same shape as x. But if the domain is real_vector, the ndim of the returned Tensor should 1 less than that of x.

The tricky part for RealNVPTransform is that now the domain can be an arbitrary shape. For example, x's spec might be rank-2. So it's difficult to define its domain shape.

Comment (emailweixu) [reply alf/networks/normalizing_flow_networks.py:532]: Then perhaps domain and codomain can be set at `__init__` as `td.constraints.independent(td.constraints.real, spec.ndim)`

Comment (hnyu) [reply alf/networks/normalizing_flow_networks.py:532]: > Then perhaps domain and codomain can be set at `__init__` as `td.constraints.independent(td.constraints.real, spec.ndim)`

Sounds good. I've made the change.

**Comment by hnyu** [comment]
After discussions with @Haichao-Zhang and @breakds , I've added a note talking about computing probabilities of transformed distributions.

File: docs/notes/compute_probs_of_transformed_dist.rst [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+    \end{array}
+
+with :math:`x=g^{-1}(g(x))`, not including :math:`\mu` in the resulting p.d.f.
+Thus PG won't be able to adjust :math:`\mu`.
+
+Transform caches
+^^^^^^^^^^^^^^^^
+
+A cache is usually helpful for a bijective transform where the inverse is computationally
+expensive or numerically instable. The way PyTorch uses a cache is that for an
```
Comment (emailweixu) [inline docs/notes/compute_probs_of_transformed_dist.rst]: instable => unstable

File: docs/notes/compute_probs_of_transformed_dist.rst [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+      - :math:`x \leftarrow x`
+      - :math:`x \leftarrow g^{-1}(g(x))` (*)
+
+'*' means potential numerical issues when inverting.
+
+A composition of transforms
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Suppose :math:`g` is now a composition of transforms
+:math:`g=g_N\circ \ldots g_k \ldots \circ g_1`, where :math:`g_k` is the first transform
```
Comment (emailweixu) [inline docs/notes/compute_probs_of_transformed_dist.rst]: where `g_k` is the *last* transform?

File: docs/notes/compute_probs_of_transformed_dist.rst [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+that contains trainable parameters. Then for PG, it is completely fine to do either
+of the following things:
+
+
+1. Simply detach :math:`y` and invalidate all caches.
+
+2. Don't detach :math:`y`, turn on caches for transforms :math:`g_N, \ldots, g_{k-1}`,
```
Comment (emailweixu) [inline docs/notes/compute_probs_of_transformed_dist.rst]: `g_N, \ldots, g_{k+1}`

File: docs/notes/compute_probs_of_transformed_dist.rst [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+   and detach the input :math:`g^{-1}_{k-1}\circ \ldots \circ g^{-1}_N(y)`
```
Comment (emailweixu) [inline docs/notes/compute_probs_of_transformed_dist.rst]: `g^{-1}_{k+1} ... `

File: docs/notes/compute_probs_of_transformed_dist.rst [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+
+In the hindsight, we don't detach the action and exploit the cache for PG to avoid
+inverting. When the transform :math:`g=g_N\circ\ldots\circ g_2` has no trainable parameters
+(e.g., ``StableTanh``), we have parameters only exist in :math:`p_{\theta}(x)`. It
+follows
+
+.. math::
+
+    \begin{array}{rcl}
+        \log p_Y(y) &=& \log p_{\theta}(g^{-1}(y)) - \log \Big|\frac{\partial g(x)}{\partial x}\big|_{x=x}\Big|\\
```
Comment (emailweixu) [inline docs/notes/compute_probs_of_transformed_dist.rst]: `\Big|\frac{\partial g(x)}{\partial x}\big|_{x=g^{-1}(y)}`

File: docs/notes/compute_probs_of_transformed_dist.rst [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+
+Action transformations as environment wrappers
+----------------------------------------------
+
+ALF's AC and PPO algorithms always detach the action from ``sample()`` for PG loss,
+without checking if the transforms have trainable parameters or not. This simplicity
+invalidate caches and sometimes causes numerical issues even when all the transforms
+do not have trainable parameters.
+
+In the hindsight, we don't detach the action and exploit the cache for PG to avoid
```
Comment (emailweixu) [inline docs/notes/compute_probs_of_transformed_dist.rst]: What does "In the hindsight" mean here?

File: docs/notes/compute_probs_of_transformed_dist.rst (hunk 11 above)
Comment (hnyu) [reply docs/notes/compute_probs_of_transformed_dist.rst]: > where `g_k` is the _last_ transform?

I added a note (from left to right).

File: docs/notes/compute_probs_of_transformed_dist.rst (hunk 15 above)
Comment (hnyu) [reply docs/notes/compute_probs_of_transformed_dist.rst]: > What does "In the hindsight" mean here?

Rephrased for clarity

**Review by emailweixu — APPROVED** [review]

**Pull request merged**